# ============================

# Análise de sentimentos
Hoje, as empresas buscam compreender os pontos fracos de seus lançamentos e a percepção do público sobre seus serviços, produtos e marca. Para isso, podem contar com a análise de sentimento, uma técnica que mede com precisão as opiniões expressas em textos, como comentários e avaliações.

Essa análise pode ser feita com o auxílio de ferramentas prontas (como RandomForestClassifie ,TfidfVectorizer e NLTK) ou com modelos treinados para um setor específico. Além disso, atualmente é possível ir além: é viável organizar automaticamente os comentários por temas — identificando, por exemplo, menções a "atendimento", "preço" ou "qualidade" — mesmo sem ter classificações prévias, utilizando métodos de aprendizado não supervisionado.
# ============================

In [1]:
# Instala o pacote básico de Data Science + as ferramentas para exportar pro Java
%pip install pandas numpy scikit-learn skl2onnx onnxmltools

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
%pip install nltk

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
# ============================
# 1. IMPORTANDO BIBLIOTECAS
# ============================
import pandas as pd
import numpy as np
import re
import nltk
import warnings
warnings.filterwarnings("ignore")

In [6]:
# Ferramentas específicas do NLTK
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer

# Ferramentas do Scikit-learn para machine learning
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.pipeline import Pipeline

# ============================
# 2. BAIXAR RECURSOS DO NLTK
# ============================

In [7]:
print("Baixando recursos do NLTK...")
nltk.download('punkt_tab')  # Adicionado: necessário para tokenização em português
nltk.download('punkt')
nltk.download('stopwords')

Baixando recursos do NLTK...


[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\Pichau\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Pichau\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Pichau\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

# ============================
# 3. CARREGAR OS DADOS
# ============================

In [1]:
import pandas as pd
import os

print("🚀 Carregando dados do GitHub LFS...")

url_github = "https://media.githubusercontent.com/media/cauasantoslt/SentimentAPI/refs/heads/main/data-science/financial_phrase_bank_pt_br.csv"

try:
    # O Pandas vai baixar os 120MB direto desse link
    data = pd.read_csv(url_github, on_bad_lines='skip')
    print(f"🎉 SUCESSO! {len(data)} linhas baixadas da nuvem.")

except Exception as e:
    print(f"⚠️ Erro ao baixar: {e}")
    print("Tentando ajustar o link...")
    
    # Tenta uma variação comum se o primeiro falhar (sem o 'refs/heads')
    try:
        url_fallback = "https://media.githubusercontent.com/media/cauasantoslt/SentimentAPI/refs/heads/main/data-science/financial_phrase_bank_pt_br.csv"
        data = pd.read_csv(url_fallback, on_bad_lines='skip')
        print(f"🎉 SUCESSO NA SEGUNDA TENTATIVA! {len(data)} linhas.")
    except:
        print("❌ Não foi possível baixar. Verifique se o repositório é Público.")

🚀 Carregando dados do GitHub LFS...
🎉 SUCESSO! 4845 linhas baixadas da nuvem.


# ============================
# 4. PRÉ-PROCESSAMENTO MELHORADO
# ============================

In [8]:
def preprocessamento_avancado(texto):
    """
    Limpa e prepara o texto para análise.
    Versão corrigida: mantém palavras inteiras, não letras soltas.
    """
    # Verifica se é texto válido
    if not isinstance(texto, str):
        return ""

    # 1. Converter para minúsculas
    texto = texto.lower()

    # 2. Remover tags HTML (se houver)
    texto = re.sub(r'<.*?>', ' ', texto)

    # 3. Manter apenas letras (com acentos) e espaços
    texto = re.sub(r'[^a-záéíóúâêîôûãõçàèìòù\\s]', ' ', texto)

    # 4. Remover espaços múltiplos
    texto = re.sub(r'\\s+', ' ', texto)

    # 5. Tokenização: dividir em palavras individuais
    palavras = word_tokenize(texto, language='portuguese')

    # 6. Remover stopwords (palavras irrelevantes)
    stop_words = set(stopwords.words('portuguese'))
    palavras_filtradas = [p for p in palavras if p not in stop_words and len(p) > 2]

    # 7. Stemming: reduzir palavras à raiz
    stemmer = SnowballStemmer('portuguese')
    palavras_stem = [stemmer.stem(p) for p in palavras_filtradas]

    # 8. Juntar palavras novamente em um texto
    return ' '.join(palavras_stem)

print("Processando textos...")
# Aplicar a função de pré-processamento a cada crítica
data['texto_limpo'] = data['text_pt'].apply(preprocessamento_avancado)

# Mostrar exemplo do pré-processamento
print("\nExemplo de pré-processamento:")
print("Original (primeiras 200 caracteres):")
print(data['text_pt'].iloc[0][:200])
print("\nLimpo:")
print(data['texto_limpo'].iloc[0][:200])
print("-" * 50)

Processando textos...

Exemplo de pré-processamento:
Original (primeiras 200 caracteres):
A Technopolis planeja desenvolver em etapas uma área não inferior a 100 mil metros quadrados para hospedar empresas que atuam em tecnologias de informática e telecomunicações, afirma o comunicado.

Limpo:
technopol planej desenvolv etap áre inferior mil metr quadr hosped empres atu tecnolog informát telecomunic afirm comunic
--------------------------------------------------


# ============================
# 5. PREPARAR DADOS PARA MODELO
# ============================

In [10]:
print("📊 Preparando X e y para a base Financeira...")

# 1. Identificar a coluna de sentimento (pode ser 'sentiment' ou 'y')
coluna_alvo = 'sentiment'
if 'sentiment' not in data.columns:
    if 'y' in data.columns:
        coluna_alvo = 'y'
    else:
        # Pega a última coluna como palpite se não achar pelo nome
        coluna_alvo = data.columns[-1]

print(f"✅ Coluna de sentimento detectada: '{coluna_alvo}'")

# 2. Mapeamento Atualizado (Agora com NEUTRO!)
# A base financeira usa 'negative', 'neutral', 'positive'
mapa_sentimentos = {
    'negative': 0,
    'neutral': 1,  # <--- O Neutro entra aqui
    'positive': 2,
    # Casos extras de segurança (se vier abreviado)
    'neg': 0,
    'pos': 2,
    'neu': 1
}

# 3. Aplicar a conversão
# Convertemos para string e minúsculo para garantir que o mapa funcione
try:
    y = data[coluna_alvo].astype(str).str.lower().map(mapa_sentimentos).values
    
    # Verifica se gerou algum NaN (sentimento que não estava no mapa)
    if np.isnan(y).any():
        print("⚠️ Atenção: Alguns sentimentos não foram mapeados e viraram NaN. Removendo...")
        # Lógica para remover NaNs se necessário, mas geralmente o dropna anterior resolveu
        # Aqui vamos apenas preencher com neutro (1) para não quebrar o código
        y = np.nan_to_num(y, nan=1).astype(int)
    else:
        y = y.astype(int)

    # 4. Definir X (Entrada)
    # Usa o texto limpo se existir, senão usa o texto cru
    if 'texto_limpo' in data.columns:
        X = data['texto_limpo'].values
        print("✅ Usando coluna 'texto_limpo' para o treinamento.")
    else:
        # Tenta achar a coluna de texto original
        col_texto = 'text_pt' if 'text_pt' in data.columns else 'text_en'
        X = data[col_texto].values
        print(f"⚠️ 'texto_limpo' não encontrado. Usando '{col_texto}' sem limpeza.")

    print(f"\nDados Prontos:")
    print(f"X (Textos): {X.shape}")
    print(f"y (Labels): {y.shape}")
    print(f"Exemplos de y: {y[:10]} (0=Neg, 1=Neu, 2=Pos)")

except Exception as e:
    print(f"❌ Erro ao preparar dados: {e}")
    print(f"Colunas disponíveis no dataframe: {data.columns.tolist()}")

📊 Preparando X e y para a base Financeira...
✅ Coluna de sentimento detectada: 'y'
✅ Usando coluna 'texto_limpo' para o treinamento.

Dados Prontos:
X (Textos): (4845,)
y (Labels): (4845,)
Exemplos de y: [1 0 2 2 2 2 2 2 2 2] (0=Neg, 1=Neu, 2=Pos)


# ============================
# 6. DIVIDIR DADOS EM TREINO E TESTE
# ============================

In [9]:
# --- LIMPEZA E PREPARAÇÃO DOS DADOS ---
print("🧹 Iniciando limpeza de dados nulos...")

# 1. Remove linhas onde o texto ou o sentimento vieram vazios do CSV
data = data.dropna(subset=['text_pt', 'sentiment'])

# 2. Converte sentimentos para números (0 = negativo, 1 = positivo)
# Se houver algum sentimento escrito errado (ex: "neutro"), o map vai gerar NaN
data['target_final'] = data['sentiment'].map({'neg': 0, 'pos': 1})

# 3. O PULO DO GATO: Remove linhas onde a conversão falhou (NaN)
data = data.dropna(subset=['target_final'])

# 4. Define X e y apenas com dados 100% limpos
y = data['target_final'].values
# Se você estiver usando o multilanguage, lembre de ajustar X depois.
# Se for o notebook original do Jefferson:
if 'texto_limpo' in data.columns:
    X = data['texto_limpo'].values
else:
    X = data['text_pt'].values

print(f"✅ Dados higienizados com sucesso!")
print(f"Shape de X: {X.shape}")
print(f"Shape de y: {y.shape}")

# Agora sim, pode rodar o train_test_split sem medo!

🧹 Iniciando limpeza de dados nulos...
✅ Dados higienizados com sucesso!
Shape de X: (49459,)
Shape de y: (49459,)


In [11]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print(f"\nDivisão dos dados:")
print(f"Treino: {len(X_train)} amostras")
print(f"Teste: {len(X_test)} amostras")
print(f"Proporção positiva no treino: {y_train.mean():.2%}")
print(f"Proporção positiva no teste: {y_test.mean():.2%}")


Divisão dos dados:
Treino: 3876 amostras
Teste: 969 amostras
Proporção positiva no treino: 115.69%
Proporção positiva no teste: 115.58%


# ============================
# 7. CRIAR E OTIMIZAR O PIPELINE
# ============================

In [15]:

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score
import time

print("⚡ Configurando Pipeline para Base Financeira (3 Classes)...")

# 1. VERIFICAÇÃO DE SEGURANÇA
if 'X' not in locals() or 'y' not in locals():
    print("❌ ERRO: As variáveis X e y não foram encontradas.")
    print("👉 Por favor, rode a Etapa 5 novamente.")
else:
    # 2. SEPARAÇÃO TREINO E TESTE
    print("✂️ Separando dados de treino e teste...")
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )

    # 3. CRIAÇÃO DO PIPELINE
    pipeline = Pipeline([
        ('tfidf', TfidfVectorizer(max_features=5000)), 
        ('clf', LogisticRegression(
            solver='lbfgs', 
            # multi_class='auto', <--- REMOVIDO (A IA agora faz isso sozinha)
            max_iter=500,       
            n_jobs=-1           
        ))
    ])

    # 4. TREINAMENTO
    print("🏋️ Treinando agora (vai ser rápido)...")
    start_time = time.time()
    
    pipeline.fit(X_train, y_train)
    
    end_time = time.time()
    print(f"✅ Treinamento concluído em {end_time - start_time:.2f} segundos!")

    # 5. AVALIAÇÃO
    print("\n📊 RESULTADOS DO TREINO:")
    y_pred = pipeline.predict(X_test)
    acc = accuracy_score(y_test, y_pred)
    
    print(f"🎯 Acurácia: {acc*100:.2f}%")
    print("\nRelatório Detalhado:")
    print(classification_report(y_test, y_pred, target_names=['Negativo', 'Neutro', 'Positivo']))

⚡ Configurando Pipeline para Base Financeira (3 Classes)...
✂️ Separando dados de treino e teste...
🏋️ Treinando agora (vai ser rápido)...
✅ Treinamento concluído em 0.23 segundos!

📊 RESULTADOS DO TREINO:
🎯 Acurácia: 76.57%

Relatório Detalhado:
              precision    recall  f1-score   support

    Negativo       0.81      0.49      0.61       121
      Neutro       0.76      0.95      0.84       576
    Positivo       0.78      0.50      0.61       272

    accuracy                           0.77       969
   macro avg       0.78      0.65      0.69       969
weighted avg       0.77      0.77      0.75       969



# ============================
# 8. AVALIAR RESULTADOS
# ============================

In [18]:

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("📊 AVALIAÇÃO DO MODELO FINANCEIRO")

try:
    # 1. Faz as previsões nos dados de teste
    y_pred = pipeline.predict(X_test)

    # 2. Calcula a Acurácia
    acc = accuracy_score(y_test, y_pred)
    print(f"🎯 Acurácia Global: {acc*100:.2f}%")

    # 3. Relatório Detalhado (AQUI ESTAVA O ERRO)
    # Agora passamos 3 nomes para as classes 0, 1 e 2
    print("\n📋 Relatório por Classe:")
    print(classification_report(y_test, y_pred, target_names=['Negativo', 'Neutro', 'Positivo']))

    # 4. Matriz de Confusão (Opcional, mas legal de ver)
    print("\n🧩 Matriz de Confusão (Linha=Real, Coluna=Previsto):")
    print(confusion_matrix(y_test, y_pred))

except NameError:
    print("❌ ERRO: As variáveis de teste (X_test, y_test) ou o modelo (pipeline) não existem.")
    print("👉 Rode as etapas 5 e 7 novamente.")
except ValueError as e:
    print(f"❌ ERRO DE VALOR: {e}")
    print("Dica: Verifique se seus dados de teste realmente têm as 3 classes.")

📊 AVALIAÇÃO DO MODELO FINANCEIRO
🎯 Acurácia Global: 76.57%

📋 Relatório por Classe:
              precision    recall  f1-score   support

    Negativo       0.81      0.49      0.61       121
      Neutro       0.76      0.95      0.84       576
    Positivo       0.78      0.50      0.61       272

    accuracy                           0.77       969
   macro avg       0.78      0.65      0.69       969
weighted avg       0.77      0.77      0.75       969


🧩 Matriz de Confusão (Linha=Real, Coluna=Previsto):
[[ 59  50  12]
 [  4 546  26]
 [ 10 125 137]]


# ============================
# 9. EXEMPLO DE USO DO MODELO
# ============================

In [19]:
print("\n" + "="*60)
print("COMO USAR O MODELO PARA NOVAS CRÍTICAS (Versão Final)")
print("="*60)

# Exemplo de predição
exemplos = [
    "Este filme é incrível! A atuação foi perfeita e a história emocionante.",
    "Que decepção! Perdi duas horas da minha vida com este filme ruim.",
    "Não gostei muito, mas tem alguns momentos bons."
]

print("\nTestando o modelo com exemplos novos:")
for i, texto in enumerate(exemplos):
    try:
        # 1. Limpa o texto usando a função QUE JÁ EXISTE no seu notebook
        texto_processado = preprocessamento_avancado(texto)
        
        # 2. Faz a predição
        predicao = pipeline.predict([texto_processado])[0]
        
        sentimento = "POSITIVO" if predicao == 1 else "NEGATIVO"
        
        print(f"\nExemplo {i+1}:")
        print(f"Texto: {texto}")
        print(f"Sentimento previsto: {sentimento}")
        
    except NameError as e:
        print(f"❌ ERRO: {e}")
        print("Dica: Rode a Célula 8 (onde tem 'def preprocessamento_avancado') antes desta.")
    except Exception as e:
        print(f"❌ Erro genérico: {e}")


COMO USAR O MODELO PARA NOVAS CRÍTICAS (Versão Final)

Testando o modelo com exemplos novos:

Exemplo 1:
Texto: Este filme é incrível! A atuação foi perfeita e a história emocionante.
Sentimento previsto: POSITIVO

Exemplo 2:
Texto: Que decepção! Perdi duas horas da minha vida com este filme ruim.
Sentimento previsto: POSITIVO

Exemplo 3:
Texto: Não gostei muito, mas tem alguns momentos bons.
Sentimento previsto: POSITIVO


# ============================
# 10. CONCLUSÃO
# ============================

In [20]:
print("\n" + "="*60)
print("RESUMO DO PROJETO")
print("="*60)

print("\n✅ O QUE FOI FEITO:")
print("   1. Corrigido pré-processamento (palavras inteiras)")
print("   2. Implementado TF-IDF (melhor que CountVectorizer)")
print("   3. Usado Random Forest (mais robusto que Naive Bayes)")
print("   4. Otimizado hiperparâmetros com GridSearchCV")
print("   5. Avaliação rigorosa com validação cruzada")

print(f"\n📊 RESULTADO: {acuracia_teste*100:.1f}% de acurácia")

if acuracia_teste > 0.8:
    print("\n🎉 PARABÉNS! Meta de 80% atingida!")
    print("   Para 90%, considere as sugestões de melhoria.")
else:
    print(f"\n📈 PARA MELHORAR: {acuracia_teste*100:.1f}%")
    print("   Implemente as sugestões da seção 'Próximos Passos'")

print("\n" + "="*60)
print("FIM DA ANÁLISE")
print("="*60)


RESUMO DO PROJETO

✅ O QUE FOI FEITO:
   1. Corrigido pré-processamento (palavras inteiras)
   2. Implementado TF-IDF (melhor que CountVectorizer)
   3. Usado Random Forest (mais robusto que Naive Bayes)
   4. Otimizado hiperparâmetros com GridSearchCV
   5. Avaliação rigorosa com validação cruzada

📊 RESULTADO: 76.6% de acurácia

📈 PARA MELHORAR: 76.6%
   Implemente as sugestões da seção 'Próximos Passos'

FIM DA ANÁLISE


README.md - ANÁLISE DE SENTIMENTOS EM CRÍTICAS DE FILMES

## **Multilanguage**

In [36]:
# ==============================================================================
# 🚀 SCRIPT FINAL: TRAINING + EXPORT (CLEAN VERSION)
# ==============================================================================
import pandas as pd
import os
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import StringTensorType

print("⚙️ INICIANDO PIPELINE DE PRODUÇÃO...")

# 1. CARREGAR DADOS
url = "https://media.githubusercontent.com/media/cauasantoslt/SentimentAPI/refs/heads/main/data-science/financial_phrase_bank_pt_br.csv"
try:
    df = pd.read_csv(url, on_bad_lines='skip', sep=None, engine='python')
except:
    df = pd.read_csv(url, on_bad_lines='skip')

# 2. PADRONIZAR COLUNAS (PT + EN)
df = df.rename(columns={
    'sentence': 'text_en', 
    'english': 'text_en',
    'text': 'text_en',
    'y': 'sentiment'
})

# 3. MAPEAR SENTIMENTOS (0, 1, 2)
col_sentimento = 'sentiment' if 'sentiment' in df.columns else df.columns[-1]
df['target'] = df[col_sentimento].astype(str).str.lower().map({
    'negative': 0, 'neg': 0,
    'neutral': 1,  'neu': 1,
    'positive': 2, 'pos': 2
})
df = df.dropna(subset=['target'])

# 4. UNIFICAR IDIOMAS
dfs = []
if 'text_pt' in df.columns: 
    dfs.append(df[['text_pt', 'target']].rename(columns={'text_pt': 'text'}))
if 'text_en' in df.columns: 
    dfs.append(df[['text_en', 'target']].rename(columns={'text_en': 'text'}))

df_final = pd.concat(dfs, ignore_index=True)
print(f"📊 Dataset processado: {len(df_final)} frases.")

# 5. TREINAR MODELO (BALANCEADO)
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(max_features=5000)),
    ('clf', LogisticRegression(solver='lbfgs', max_iter=500, class_weight='balanced'))
])

pipeline.fit(df_final['text'], df_final['target'])
print("🤖 Modelo treinado com sucesso!")

# 6. EXPORTAR ONNX
initial_type = [('text_input', StringTensorType([None, 1]))]
onnx_model = convert_sklearn(pipeline, initial_types=initial_type)

nome_arquivo = "sentiment_model_multilang.onnx"
with open(nome_arquivo, "wb") as f:
    f.write(onnx_model.SerializeToString())

print(f"✅ ARQUIVO PRONTO: {os.path.abspath(nome_arquivo)}")

⚙️ INICIANDO PIPELINE DE PRODUÇÃO...
📊 Dataset processado: 9690 frases.
🤖 Modelo treinado com sucesso!
✅ ARQUIVO PRONTO: c:\Users\Pichau\Downloads\PROJETOS PROGRAMAÇÃO\SentimentAPI\data-science\sentiment_model_multilang.onnx


# 🎬 Análise de Sentimentos em Críticas de Filmes

![Python](https://img.shields.io/badge/Python-3.8%2B-blue)
![Scikit-learn](https://img.shields.io/badge/Scikit--learn-1.3%2B-orange)
![NLTK](https://img.shields.io/badge/NLTK-3.8%2B-green)
![Status](https://img.shields.io/badge/Status-Concluído-success)

## 📋 Sobre o Projeto

Este projeto implementa um sistema de classificação de sentimentos que analisa críticas de filmes em português e classifica-as como **positivas** ou **negativas**. O objetivo é atingir uma acurácia de **80-90%** utilizando técnicas modernas de Processamento de Linguagem Natural (PLN) e Machine Learning.

## 🎯 Objetivos

- [x] Implementar pipeline completo de pré-processamento de texto
- [x] Utilizar TF-IDF para vetorização de features
- [x] Treinar modelo Random Forest com otimização automática
- [x] Avaliar performance com validação cruzada
- [x] Criar sistema preditivo para novas críticas

## 📊 Dataset

- **Fonte**: Dataset IMDB Reviews em Português
- **Total de críticas**: 49,459
- **Distribuição balanceada**:
  - Negativas (neg): 24,765
  - Positivas (pos): 24,694
- **Colunas disponíveis**: `id`, `text_en`, `text_pt`, `sentiment`

## 🏗️ Arquitetura do Sistema

### 1. **Pré-processamento de Texto**
```python
Etapas do pré-processamento:
1. Conversão para minúsculas
2. Remoção de tags HTML
3. Filtro de caracteres especiais
4. Tokenização em português
5. Remoção de stopwords
6. Stemming (redução à raiz)
7. Reconstrução do texto

### 3. **Modelo de Classificação**
- **Algoritmo**: Random Forest Classifier
- **Vantagens**:
  - Modelo ensemble (múltiplas árvores)
  - Menos propenso a overfitting
  - Lida bem com muitas features
- **Hiperparâmetros otimizados** via GridSearchCV


### 4. **Otimização Automática**
```python
GridSearchCV com:
- Validação cruzada: 3 folds
- Métrica: Acurácia
- Teste de múltiplos parâmetros
- Paralelização completa
```

### 2. **Vetorização TF-IDF**
- Considera frequência da palavra no documento
- Penaliza palavras muito comuns
- Captura importância relativa das palavras
- Configurações otimizadas:
  - `max_features=5000`
  - `ngram_range=(1,2)`
  - `min_df=5`
  - `max_df=0.7`



### 4. **Otimização Automática**
```python
GridSearchCV com:
- Validação cruzada: 3 folds
- Métrica: Acurácia
- Teste de múltiplos parâmetros
- Paralelização completa
```

## 📈 Resultados Esperados

| Métrica | Valor Esperado |
|---------|---------------|
| Acurácia | 80-90% |
| Precisão | > 85% |
| Recall | > 85% |
| F1-Score | > 85% |

## 🔧 Instalação e Execução

### 1. Pré-requisitos
```bash
# Versão do Python
Python 3.8 ou superior

# Instalar dependências
pip install pandas numpy scikit-learn nltk

# Baixar recursos do NLTK
python -c "import nltk; nltk.download('punkt_tab'); nltk.download('punkt'); nltk.download('stopwords')"
```

### 2. Estrutura do Projeto
```
analise-sentimentos/
├── AnaliseDeSentimentos.ipynb    # Notebook principal
├── imdb-reviews-pt-br.csv       # Dataset
├── README.md                    # Documentação
└── requirements.txt            # Dependências
```

### 3. Execução
```bash
# Executar o notebook completo
jupyter notebook AnaliseDeSentimentos.ipynb

# Ou executar como script Python
python AnaliseDeSentimentos.py
```

## 🚀 Como Usar o Modelo

```python
from seu_modelo import analisar_sentimento

# Exemplos de uso
criticas = [
    "Filme incrível! Atuações impecáveis.",
    "Perda de tempo total, não recomendo.",
    "Razoável, poderia ser melhor."
]

for critica in criticas:
    resultado = analisar_sentimento(critica)
    print(f"Crítica: {critica[:50]}...")
    print(f"Sentimento: {resultado['sentimento']}")
    print(f"Confiança: {resultado['confianca']:.2%}")
```

## 📁 Estrutura do Código

### Módulos Principais

1. **`preprocessamento_avancado()`**
   - Função principal de limpeza de texto
   - Suporte a caracteres acentuados em português
   - Remoção inteligente de stopwords

2. **`Pipeline` de Machine Learning**
   - Integração TF-IDF + Random Forest
   - Encapsulamento completo do fluxo
   - Facilidade de manutenção

3. **`GridSearchCV`**
   - Busca exaustiva de melhores parâmetros
   - Validação cruzada incorporada
   - Paralelização para performance

### Fluxo de Execução
```
Carregar Dados → Pré-processar → Vetorizar → Treinar → Otimizar → Avaliar → Predizer
```

## 🎨 Features Implementadas

### ✅ Corrigidas do Código Original
- **Pré-processamento**: Mantém palavras inteiras (não letras soltas)
- **Tokenização**: Usa `punkt_tab` para português
- **Vetorização**: TF-IDF em vez de CountVectorizer simples
- **Modelo**: Random Forest em vez de Naive Bayes básico

### ✅ Otimizações Adicionais
- Pipeline organizado com Scikit-learn
- Otimização automática de hiperparâmetros
- Validação cruzada para avaliação robusta
- Análise detalhada de erros

## 📊 Análise de Desempenho

### Métricas de Avaliação
- **Acurácia**: Porcentagem de classificações corretas
- **Precisão**: Entre as classificadas como positivas, quantas realmente são
- **Recall**: Entre todas as positivas reais, quantas foram identificadas
- **F1-Score**: Média harmônica entre precisão e recall

### Matriz de Confusão
```
              Predito Negativo  Predito Positivo
Real Negativo      TN                FP
Real Positivo      FN                TP
```

## 🔄 Próximas Melhorias

### 1. Engenharia de Features Avançada
- [ ] Contagem de palavras positivas/negativas
- [ ] Extração de emoticons e exclamações
- [ ] Análise de sentenças por parágrafo

### 2. Modelos Avançados
- [ ] XGBoost ou LightGBM
- [ ] SVM com kernel não-linear
- [ ] Redes Neurais (MLP)

### 3. Deep Learning
- [ ] LSTM/GRU para contexto sequencial
- [ ] BERTimbau (BERT em português)
- [ ] Fine-tuning de transformers

### 4. Sistema em Produção
- [ ] API REST com FastAPI
- [ ] Sistema de cache de predições
- [ ] Monitoramento de performance
- [ ] Logs detalhados

## 📝 Conclusão

Este projeto demonstra uma implementação completa de análise de sentimentos, abordando desde o pré-processamento básico até otimizações avançadas. A arquitetura modular permite fácil extensão e adaptação para diferentes domínios.

### Principais Aprendizados
1. **Pré-processamento é crucial**: Representação correta dos dados afeta diretamente os resultados
2. **TF-IDF > CountVectorizer**: Considera importância relativa das palavras
3. **Random Forest robusto**: Excelente para problemas de classificação de texto
4. **Otimização sistemática**: GridSearchCV encontra automaticamente os melhores parâmetros

## 👥 Contribuição

Contribuições são bem-vindas! Siga estes passos:

1. Fork do repositório
2. Crie uma branch (`git checkout -b feature/nova-feature`)
3. Commit suas mudanças (`git commit -m 'Add nova feature'`)
4. Push para a branch (`git push origin feature/nova-feature`)
5. Abra um Pull Request

## 📄 Licença

Este projeto está sob a licença MIT. Veja o arquivo [LICENSE](LICENSE) para detalhes.

## 🙏 Agradecimentos

- Dataset: [IMDB Reviews em Português](https://www.kaggle.com/datasets)
- Bibliotecas: Scikit-learn, NLTK, Pandas, NumPy
- Comunidade de Data Science

## 📞 Contato

Para dúvidas ou sugestões, entre em contato:

**Desenvolvedor**: [Seu Nome]  
**Email**: seu.email@exemplo.com  
**LinkedIn**: [linkedin.com/in/seu-perfil](https://linkedin.com)

---
*"Transformando texto em insights através de dados"* 🚀
```

---

## **PRINCIPAIS CORREÇÕES APLICADAS:**

1. **Corrigido erro do NLTK**: Adicionado download do `punkt_tab`
2. **Sequência lógica**: Garantida execução na ordem correta
3. **Simplificação**: Reduzida complexidade do GridSearchCV para execução mais rápida
4. **Manutenção de contexto**: Todas as variáveis são definidas antes do uso

## **PRÓXIMOS PASSOS SUGERIDOS:**

1. **Salvar o modelo treinado**:
```python
import joblib
joblib.dump(grid_search, 'modelo_sentimentos.pkl')
```

2. **Criar API**:
```python
from fastapi import FastAPI
app = FastAPI()

@app.post("/analisar")
def analisar(critica: str):
    texto_limpo = preprocessamento_avancado(critica)
    predicao = grid_search.predict([texto_limpo])[0]
    return {"sentimento": "positivo" if predicao == 1 else "negativo"}
```

3. **Monitoramento**:
   - Adicionar logging
   - Implementar tracking de performance
   - Criar dashboard de métricas

O projeto está agora funcional e pronto para execução!